### PR TITLE
[Test] Prefix-cache passthrough test coverage prerequisite for W2 (#4855)

### PR DESCRIPTION
**Pull request by ShengleiFu**
## Purpose

Add characterization tests for the prefix-cache passthrough D2H path (`build_mm_cpu`/`get_merged_multimodal_states` in `prefix_cache.py`/`mm_outputs.py`) ahead of **W2** in #4855's self-review checklist, per its own "worker test coverage" caution before touching anything beyond W1.

This PR does **not** change runtime behavior; it locks down current passthrough/D2H semantics so the W2 refactor (batching per-key D2H, extending the `gpu_resident_buffer_keys` mechanism, or a same-GPU zero-copy path) has a regression safety net, regardless of which approach gets picked.

This PR:
- Pins the per-request D2H cost of `build_mm_cpu`: the tensor-copy count grows with batch size through `codes.ref`, so W2's batching is what moves it. Counting tensor leaves rather than dict keys is deliberate — for a flat dict the count is true by construction (`tests/utils/test_mm_outputs.py`)
- Asserts `build_mm_cpu` returns CPU tensors for every nested leaf, with a GPU-gated variant. RFC #4855's "skip the conversion when the consumer is the same-GPU next stage" option is invisible to a CPU-only suite, where `.to("cpu")` is a no-op
- Covers the passthrough shape qwen3-tts actually emits: a nested `codes`/`meta` dict holding a per-request tensor list (`codes.ref`) plus int32/int8 token-aligned metadata, through the `pass_lists_through` branch. The pre-existing list coverage used a list of strings, and no test carried an int dtype
- Covers a passthrough tensor that is **not** token-aligned (`to_payload_element`'s `element.clone()` branch) — a W2 batcher assuming everything is token-aligned is the most likely regression, and it had no coverage
- Covers `build_mm_cpu`/`_to_cpu`'s nested dict/list recursion, non-contiguous→contiguous, detach, and dtype/shape/value preservation
- Adds `discard_deferred_mm_outputs` (abort path) and request-ID-reuse-after-discard coverage — both previously zero test coverage (`tests/core/test_prefix_cache.py`)
- Covers a real passthrough tensor (1D token metadata + 2D feature tensor) sliced per request under mixed cache hit/miss + uneven scheduled length, extending `test_get_merged_multimodal_outputs`
- Adds the #4106 async-write `schedule_async_write`/`drain_ready_async_writes` lifecycle test (`tests/core/test_prefix_cache_async_write.py`)
- Fixes that file's platform mark: it was `cuda`, and no Buildkite step selects `cuda` for `tests/core`, so all 116 lines were collected by nothing. The two steps that reach `tests/core` select `core_model and cpu` and run on `l4_1`, so `cpu` is the lane selector and the existing `skipif` remains the hardware gate
- Covers the worker's prefix-cache branch: `_build_omni_mm_payload` selecting `_build_combined_prefix_cache_mm_payload` over the `build_mm_cpu` fallback, and `_unwrap_lists` resolving a list-valued key by batch index. The merge itself is stubbed in this test; it pins branch selection and index mapping, not the merge (`tests/worker/test_gpu_ar_model_runner.py`)

Not a duplicate: no open PR touches `test_prefix_cache.py` / `test_gpu_ar_model_runner.py` / `test_mm_outputs.py` for W2 test coverage as of this PR.

## Test Plan

**vLLM Version:** 0.26.0
**Base:** rebased onto `67c54777` (origin/main)

- [x] `tests/utils/test_mm_outputs.py`
  - [x] per-request D2H growth, parametrized over batch size
  - [x] int32/int8 dtype preservation
  - [x] CPU device on every nested leaf, plus a GPU-gated variant
  - [x] nested dict/list, non-contiguous, detach, dtype/shape/value semantics
- [x] `tests/core/test_prefix_cache.py`
  - [x] `discard_deferred_mm_outputs` abort path
  - [x] request-ID reuse after discard
  - [x] passthrough tensor slicing, 1D + 2D, mixed cache hit/miss + uneven length
  - [x] per-request tensor list passed through whole, cloned per request
  - [x] non-token-aligned passthrough broadcast, parametrized over three shapes
- [x] `tests/core/test_prefix_cache_async_write.py`
  - [x] drain-not-ready / drain-once-ready lifecycle
  - [x] consecutive `schedule_async_write` doesn't drop the previous pending write
  - [x] collected by `-m 'core_model and cpu'` (0 before this PR, 2 after)
- [x] `tests/worker/test_gpu_ar_model_runner.py`
  - [x] `build_mm_cpu` fallback path, partial downstream batch
  - [x] `_build_combined_prefix_cache_mm_payload` branch selection, partial downstream batch
  - [x] `_unwrap_lists` batch-index selection for a list-valued key

## Test Result

Run on an RTX PRO 6000 Blackwell, vLLM 0.26.0 / torch 2.11.0+cu130 / Python 3.12, in a clean worktree at the pushed commit:

```
pytest tests/core/test_prefix_cache.py tests/core/test_prefix_cache_async_write.py \
       tests/utils/test_mm_outputs.py tests/worker/test_gpu_ar_model_runner.py
64 passed
```

Nothing skipped — the CUDA-gated tests executed rather than skipping. The same four files under the CI selector (`-m 'core_model and cpu'`) select all 64 and pass.

Collection count for `tests/core/test_prefix_cache_async_write.py` under that selector: **0 before this PR, 2 after**.

Each new test was checked against a deliberately broken source tree to confirm it can fail:

| mutation | new test | rest of suite |
| --- | --- | --- |
| drop the token-aligned guard, always slice | fails | passes |
| index `codes.ref` here instead of passing it through whole | fails | passes |
| stop recursing into lists in `_to_cpu` | fails | passes |
| stage passthrough through a float buffer | fails | passes |
| `_unwrap_lists` broadcasting `v[0]` instead of `v[idx]` | fails | passes |
| skip the CPU conversion (RFC W2's same-GPU option) | fails | passes |

The last row is the one worth noting: with that change applied, every other test in these files still passes.

- `ruff check` / `ruff format` clean (pinned v0.14.10)
- `tools/pre_commit/check_test_marks.py` passes on all touched files

AI assistance used (Claude); reviewed every changed line and ran the tests above myself.

### CODE REVIEW

File: tests/core/test_prefix_cache_async_write.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
+import pytest
+import torch
+
+from vllm_omni.core.prefix_cache import OmniTensorPrefixCache
+
+pytestmark = [
+    pytest.mark.core_model,
+    pytest.mark.cuda,
+    pytest.mark.skipif(not torch.cuda.is_available(), reason="requires a CUDA device"),
+]
```
Comment (linyueqian) [inline tests/core/test_prefix_cache_async_write.py:13]: This whole file runs in **no CI lane**, so its 116 lines never execute.

The only Buildkite step that reaches `tests/core` is `Simple · Other Test`, which selects `-m 'core_model and cpu'` (`.buildkite/cuda/test-ready.yml:27`). Every `core_model and cuda` step is path-scoped elsewhere — `tests/diffusion/*`, `tests/distributed/`, `tests/engine/`, `tests/entrypoints/`, `tests/model_executor/`, `tests/model_tests/diffusion`. None of them include `tests/core`. So `cuda` here means "collected by nothing".

The convention that works is already in the sibling file: `tests/core/test_prefix_cache.py` keeps a module-level `cpu` mark and gates individual GPU tests with `@pytest.mark.skipif(not torch.cuda.is_available(), ...)`, so the cpu lane collects them and they skip cleanly on non-GPU runners.

Suggest matching that — `cpu` instead of `cuda`, keeping the `skipif` you already have.

File: tests/utils/test_mm_outputs.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    spy = mocker.spy(mm_outputs_mod, "_to_cpu")
+
+    multimodal_outputs = {
+        "key_a": torch.rand(8, 4),
+        "key_b": torch.rand(8, 4),
+        "key_c": torch.rand(8, 4),
+    }
+    result = build_mm_cpu(multimodal_outputs=multimodal_outputs)
+
+    assert spy.call_count == 3
```
Comment (linyueqian) [inline tests/utils/test_mm_outputs.py]: Two things about this baseline.

First, with a flat dict of three tensors this assertion is true by construction: `build_mm_cpu` is `for k, v in multimodal_outputs.items(): _to_cpu(v)` (`vllm_omni/utils/mm_outputs.py:102-103`), so the count *is* `len(dict)`. It can only fail if someone rewrites the loop.

Second, and more importantly — the docstring says W2's batching must *reduce* this count, which means the test is expected to fail when W2 lands. A regression net you plan to delete during the change it guards isn't guarding much.

The version of this that would hold: use a **list-of-tensors** payload, where the count is `N_requests` rather than `len(dict)`. That's the real shape — `_to_cpu` recurses into lists one D2H at a time (`mm_outputs.py:120-123`), and the qwen3-tts talker emits exactly that as `codes.ref` (`qwen3_tts_talker.py:633`, one tensor per request). That's the cost the RFC is pointing at, and a batched W2 would legitimately change the count in a direction you could assert.

File: tests/utils/test_mm_outputs.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    assert inner.shape == non_contig.shape
+    assert torch.equal(inner, non_contig)
+    assert result["nested"]["inner_scalar"] == "keep-me"
+
+    lst = result["list_of_tensors"]
+    assert torch.equal(lst[0], torch.tensor([1.0, 2.0]))
+    assert lst[1] == "marker"
+    assert torch.equal(lst[2], torch.tensor([3.0, 4.0]))
+
+    assert result["grad_key"].requires_grad is False
```
Comment (linyueqian) [inline tests/utils/test_mm_outputs.py:144]: Nothing in this PR asserts the **device** of anything `build_mm_cpu` returns — I grepped the diff; all eight `device` occurrences are `device="cuda"` construction kwargs or the skipif reason.

That matters more than it sounds, because D2H is the entire purpose of the function, and on the cpu lane `.to("cpu")` is a silent no-op. RFC #4855's W2 lists three options, one of which is *"skip CPU conversion when the consumer is the same-GPU next stage"* — that option would leave tensors on GPU and pass every test in this PR.

A `.device.type == "cpu"` assertion here would close that, and a GPU-gated variant would close it properly.

**Review by linyueqian — COMMENTED** [review]
Sorry this sat for 11 days — that's on me, not on you. Worth noting the automated reviewer never ran either: the Codex bot posted a billing-limit message two minutes after you opened it and nothing else has looked at this since.

Thanks for doing the unglamorous half first. `discard_deferred_mm_outputs` genuinely had zero coverage before this, and the slicing test at `tests/core/test_prefix_cache.py:705` is the real thing — it drives `get_merged_multimodal_states` with a live cache, mixed hit/miss, and uneven scheduled lengths, and asserts exact per-request slices. A W2 that misaligned request boundaries would fail it. That one test does the job it claims.

Where I'd push back is on the PR as a **W2 safety net** specifically. I went looking for what W2 could plausibly change and asked which of those a reviewer would catch here, and the honest answer is: mostly that one test. Three gaps stand out, all additive to fix:

**1. Device is never asserted.** Details inline. D2H is the whole point of `build_mm_cpu`, and one of the RFC's own three W2 options — leaving passthrough on GPU for a same-GPU consumer — would pass this suite silently.

**2. The real payload shape isn't exercised.** The tests use flat float32 tensors. The passthrough set actually reaching `build_mm_cpu` each decode step is `codes.ref` (a **list of per-request tensors**), plus 1-D `meta.ref_code_len` (int32) and `meta.codec_streaming` — see `qwen3_tts_talker.py:627-637`, and `prefix_cache.py:103-108` for why only 2-D `shape[0] == seq_len` tensors become cache keys and everything else falls through as passthrough. The list case is the one the RFC calls out as the cost driver (`_to_cpu` recursing per request), and it goes through `pass_lists_through=True` (`prefix_cache.py:597-604`), a branch currently covered only by a list of *strings* at `tests/core/test_prefix_cache.py:551`. No int dtype appears anywhere in the diff either.

**3. A quarter of the added lines run in no lane.** `test_prefix_cache_async_write.py` is `cuda`-marked, and no Buildkite step runs `tests/core` with a cuda selector. Details inline. This one is a marking fix, not a test-quality issue.

Two smaller things, take or leave:

- `test_build_omni_output_uses_combined_prefix_cache_mm_payload_for_partial_downstream_batch` stubs `_prepare_prefix_cache_pooler_payload_sources`, so `get_merged_multimodal_states` / `build_mm_cpu` / `_to_cpu` never run — it exercises the branch selection at `gpu_ar_model_runner.py:897-902` rather than the merge itself. Worth softening the claim in the PR body. Its `_unwrap_lists` helper (`:871-876`) also stays unexercised because the fixture has no list values, which is the same gap as (2).
- `assert not torch.equal(output.inter_stage_outputs[2]["codes.audio"], codes[1:2])` can't fail once the preceding equality passed — `codes[1:2]` and `codes[2:3]` are distinct constants.

**Separate finding, not this PR's job but relevant to W2's design:** `to_payload_element` with `pass_lists_through=True` clones the *entire* list for *every* request (`mm_outputs.py:183-186`), and `_unwrap_lists` then indexes one element out. That's O(N²) CPU clones per step for `codes.ref`. It may be a real slice of the ~40% host-sample figure in the RFC and looks like a natural target alongside the D2H batching — worth measuring before you settle on W2's approach.

**Minimum to unblock, in order:**

1. Rebase — currently conflicting. `tests/core/test_prefix_cache.py` now has the identical `pytestmark` on main, so that hunk is redundant; the worker file collides on an append after the vLLM 0.26 rebase.
2. Fix the async-write file's marks so it runs somewhere.
3. Add a list-of-tensors passthrough case through `get_merged_multimodal_states`, modelled on `codes.ref`.
4. Assert device on `build_mm_cpu` output.
5. Add a passthrough tensor with `shape[0] != total_scheduled_tokens` (the clone branch at `mm_outputs.py:191-194`) — a batcher assuming everything is token-aligned is the most likely W2 bug and has no test today.

(3)–(5) are the ones I'd actually insist on; they're what make this a net rather than a snapshot. Happy to look again quickly once it's rebased — I don't want this blocking W2 for another two weeks.

**Comment by ShengleiFu** [comment]
Hi @linyueqian, thanks for the review, I just addressed all your feedback above, plz take another look, thanks!

**Review by linyueqian — APPROVED** [review]
That was fast — thank you, and you went past what I asked for.

Verified on an L20X against your head `08753adc`:

- `tests/utils/test_mm_outputs.py`, `tests/core/test_prefix_cache.py`, `tests/core/test_prefix_cache_async_write.py`, `tests/worker/test_gpu_ar_model_runner.py` → **64 passed**
- `pytest tests/core/test_prefix_cache_async_write.py -m "core_model and cpu"` → **2 passed**, where it collected zero before. The marker fix works.

Checking each item off:

- **Device** — `test_build_mm_cpu_returns_cpu_tensors_for_every_nested_leaf` and `test_build_mm_cpu_moves_a_gpu_payload_off_the_device`. Asserting the source is on `cuda` before the call is the part that makes it a real test rather than a tautology on a CPU runner. This is the one that would have caught "leave passthrough on GPU for a same-GPU consumer" passing silently.
- **List passthrough** — `test_get_merged_multimodal_outputs_passes_per_request_tensor_list_through`, and `_unwrap_lists`' batch-index selection in the worker test. That's the `codes.ref` path, which was the largest gap.
- **Non-token-aligned** — `test_get_merged_multimodal_outputs_broadcasts_non_token_aligned_passthrough`, parametrized. This pins the clone/broadcast branch a naive `torch.cat` batcher would corrupt.
- **Int dtypes** — `test_build_mm_cpu_preserves_int_dtypes_of_talker_metadata`, matching the real `meta.ref_code_len`.
- **The call-count reframe** — `test_build_mm_cpu_d2h_count_grows_per_request`, parametrized on request count. This is the version that survives W2 instead of being deleted by it, which was the point. Nice.

LGTM. Approving so W2 isn't waiting on this any longer.

Two things for when you start W2, neither blocking here:

The O(N²) clone I mentioned is worth measuring before you pick an approach — `to_payload_element` with `pass_lists_through=True` clones the whole list per request (`mm_outputs.py:183-186`) and `_unwrap_lists` then indexes one element out. For `codes.ref` at N requests that's N² clones per step, and it may be a meaningful share of the ~40% host-sample figure. It could turn out to be a bigger lever than the D2H batching itself.

Also worth reading the W3 thread before you scope W2: https://github.com/vllm-project/vllm-omni/issues/4855#issuecomment-5159127254 — `build_mm_cpu` already hoists the D2H out of the per-request path, which is exactly the mechanism W2 is built on, and I got the neighbouring W3 row wrong by not accounting for it. Worth confirming the W2 premise against the code the same way before you commit to a design; I'd rather you catch me twice than build on another bad row.

**Comment by ShengleiFu** [comment]
Hi @linyueqian , could you help add ready label, thanks for the approval and the guidance for rest of the work in W2

**Comment by linyueqian** [comment]
@ShengleiFu can you ping me after this version's release? we are freezing every code changes for now.

**Comment by ShengleiFu** [comment]
> @ShengleiFu can you ping me after this version's release? we are freezing every code changes for now.

Will do, thanks!

**Comment by ShengleiFu** [comment]
Hi @linyueqian , version release is done, could you help add `ready` label, thanks!

**Comment by ShengleiFu** [comment]
Hi @linyueqian , sorry to bother you again, could you help re-add `ready` label(I accidentally merged main yesterday), or could you help merge this pr if the previous CI run looks good to you(infra reason caused amd run hang), thanks!

**Comment by ShengleiFu** [comment]
Hi @linyueqian , thanks for the patience, seems both issues in CI are not related to my change, could you check and see if we could merge this pr, thanks!
